src/bucket/BucketManagerImpl.cpp

```diff
@@ -867,4 +868,165 @@ BucketManagerImpl::isShutdown() const
+    std::vector<Hash> hashOrderToVisit;
+    for (uint32_t i = BucketList::kNumLevels; i > 0; --i)
+    {
+        HistoryStateBucket const& hsb = has.currentBuckets.at(i - 1);
+        Hash snap = hexToBin256(hsb.snap);
+        Hash curr = hexToBin256(hsb.curr);
+        hashesInBl.insert(snap);
+        hashesInBl.insert(curr);
+        std::vector<Hash> expected = {snap, curr};
+        if (hsb.next.hasOutputHash())
+        {
+            Hash next = hexToBin256(hsb.next.getOutputHash());
+            hashesInFuture.insert(next);
+            expected.emplace_back(next);
+        }
+        for (auto const& e : expected)
+        {
+            if (isZero(e))
+            {
+                continue;
+            }
+            hashOrderToVisit.emplace_back(e);
+            if (hashesInDir.find(e) == hashesInDir.end())
+            {
+                CLOG_ERROR(Bucket,
+                           "Missing expected level {} file for bucket {}",
+                           i - 1, binToHex(e));
+                ok = false;
+            }
+            else
+            {
+                CLOG_INFO(Bucket, "Found expected level {} file for bucket {}",
+                          i - 1, binToHex(e));
+            }
+        }
+    }
+
+    for (auto const& h : hashesInDir)
+    {
+        if (hashesInBl.find(h) == hashesInBl.end() &&
+            hashesInFuture.find(h) == hashesInFuture.end())
+        {
+            // This is a hash not-in the BL or a FutureBucket. Warn, but not
+            // error.
+            CLOG_WARNING(Bucket, "found unexpected bucket {}", binToHex(h));
+            hashOrderToVisit.emplace_back(h);
+        }
+    }
+
+    // Then we scan the existing buckets, collecting entries from those that are
+    // relevant to the current ledger state into the provided ledger map.
+    //
+    // We do this in 3 passes so we can get relatively-useful isolated IO and
+    // hashing performance metrics out of the first 2 passes (to get a hint if
+    // the user's machine is slow in some important way).
+    for (auto h : hashOrderToVisit)
+    {
+        auto path = bucketFilename(h);
+        size_t sz = fs::size(path);
+
+        auto b = std::make_shared<Bucket>(path, h);
+
+        CLOG_INFO(Bucket, "Scaning bucket {}", binToHex(h));
+        {
+            // First pass measures IO rate and checks XDR decoding.
+            BucketInputIterator in(b);
+            try
+            {
+                scanBucket("Decoded", sz, in);
+            }
+            catch (FileSystemException& e)
+            {
+                // This ERROR-logs at the XDRInputFileStream level already.
+                ok = false;
+            }
+            catch (xdr::xdr_runtime_error& e)
+            {
+                CLOG_ERROR(Bucket, "XDR error decoding {}: {}", path, e.what());
+                ok = false;
+            }
+        }
+        {
+            // Second pass measures CPU hash rate and checks hash value.
```

Comment:
That doesn't benchmark IO or CPU. I could modify it to do so but I'm assuming it wouldn't be appreciated to do two passes over the buckets while validating, which means it'd only benchmark CPU and I'd still need a separate scan loop to benchmark IO. Felt easier to just reuse one scan loop locally (and synchronously).

Reply:
Oh my thinking was that we could tweak VerifyBucketWork a bit and make hashing optional. Then in here you could just execute VerifyBucketWork several times with the parameters tuned for each use case (so it would basically replace `scanBucket`). Otherwise, we have quite a bit of duplication, as `VerifyBucketWork` already scans and hashes the bucket, checks the hash and logs the result. What do you think?

Reply:
It's possible. I remain hesitant:
  - We still need something like the `scanBucket` pass in order to build up the ledger map in memory that we later use in `LedgerManagerImpl::selfCheck` to verify the DB ledger contents.
  - We can't just "pass a closure in that collects elements" into `VerifyBucketWork` like we're doing in the synchronous `scanBucket` passes here; `VerifyBucketWork` runs on background threads and verifies each bucket in parallel rather than sequentially, and we need to sequence the bucket scans oldest-to-newest to overwrite old LEs with new ones while scanning. That's the point of the explicit `hashOrderToVisit` vector.
  - I _could_ make a dedicated `Work` subclass -- merging part of the `BucketManagerImpl::selfCheck` code and the `LedgerManagerImpl::selfCheck` -- that does a 2 phase "build up in-memory ledger map" + "verify against DB" sequence, if you've a strong preference for my (re)using `Work` classes. It feels a bit like redoing the whole PR but I guess I can try (the one major advantage to that is that it at least catches ctrl-C and gracefully exits!)

Reply:
Actually now that I browse a bit more, it looks like maybe I can reuse the BucketListIsConsistentWithDatabase invariant, which would cut down on new/redundant code here quite a bit. I'll try that.

Reply:
Sigh, no, unfortunately that only works during (old-to-new) application when there are no LEs newer than the bucket just-applied.